File: frontend/src/api/waitlist.ts
import { useMutation } from '@tanstack/react-query'
import { api } from './client'

interface WaitlistResponse {
  message: string
  already_registered: boolean
}

export function useJoinWaitlist() {
  return useMutation({
    mutationFn: async (email: string): Promise<WaitlistResponse> => {
      const { data } = await api.post('/waitlist', { email })
      return data
    },
  })
}

File: backend/app/routers/waitlist.py
from fastapi import APIRouter, Depends
from sqlalchemy import select
from sqlalchemy.ext.asyncio import AsyncSession

from app.db import get_db
from app.models.waitlist import WaitlistEntry
from app.schemas.waitlist import WaitlistRequest, WaitlistResponse

router = APIRouter(prefix="/api/waitlist", tags=["waitlist"])


@router.post("", response_model=WaitlistResponse)
async def join_waitlist(body: WaitlistRequest, db: AsyncSession = Depends(get_db)):
    existing = (await db.execute(
        select(WaitlistEntry).where(WaitlistEntry.email == body.email)
    )).scalar_one_or_none()

    if existing:
        return WaitlistResponse(message="You're already on the list!", already_registered=True)

    entry = WaitlistEntry(email=body.email)
    db.add(entry)
    await db.commit()
    return WaitlistResponse(message="You're on the list!")
